test/freenet/support/math/BootstrappingDecayingRunningAverageTest.java

```diff
@@ -0,0 +1,93 @@
+package freenet.support.math;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasEntry;
+
+import freenet.support.SimpleFieldSet;
+import org.junit.Test;
+
+public class BootstrappingDecayingRunningAverageTest {
+    @Test
+    public void decaysOverSubsequentReports() {
+        BootstrappingDecayingRunningAverage average = new BootstrappingDecayingRunningAverage(0, 0, 1, 2, null);
+        average.report(1);
+        assertThat(average.currentValue(), equalTo(1.0));
+        average.report(0);
+        assertThat(average.currentValue(), equalTo(0.5));
+        average.report(0);
+        assertThat(average.currentValue(), equalTo(0.25));
+        average.report(0);
+        assertThat(average.currentValue(), equalTo(0.125));
+    }
+
+    @Test
+    public void newInstanceHasDefaultValue() {
+        BootstrappingDecayingRunningAverage average = new BootstrappingDecayingRunningAverage(1, 0, 1, 2, null);
+        assertThat(average.currentValue(), equalTo(1.0));
+    }
+
+    @Test
+    public void countsNumberOfValidReports() {
+        BootstrappingDecayingRunningAverage average = new BootstrappingDecayingRunningAverage(0, 0, 1, 2, null);
+        assertThat(average.countReports(), equalTo(0L));
+
+        // Valid report
+        average.report(0);
+        assertThat(average.countReports(), equalTo(1L));
+
+        // Invalid reports
+        average.report(-1000);
+        average.report(1000);
+        average.report(Double.NEGATIVE_INFINITY);
+        average.report(Double.POSITIVE_INFINITY);
+        average.report(Double.NaN);
+        assertThat(average.countReports(), equalTo(1L));
```

Comment:
also check that the value is 0?

Reply:
Fair enough, added an assertion for that!